ostree/ostree_dest.go

```diff
@@ -432,13 +516,34 @@ func (d *ostreeImageDestination) Commit(ctx context.Context) error {
 
 	var selinuxHnd *C.struct_selabel_handle
 
-	if os.Getuid() == 0 && selinux.GetEnabled() {
-		selinuxHnd, err = C.selabel_open(C.SELABEL_CTX_FILE, nil, 0)
-		if selinuxHnd == nil {
-			return errors.Wrapf(err, "cannot open the SELinux DB")
-		}
+	// dlopen.GetHandle will fail if libselinux.so is not available
+	// following block will be completely jump over and
+	// selinuxHnd = nil (important in fixFiles)
+	// so the absence of SElinux will be simply ignored
+	libselinuxHnd, err := dlopen.GetHandle([]string{"libselinux.so"})
```

Comment:
This must use a versioned soname.

Also, are we comfortable with making this a silent fail-open situation? I’m not really, I’d much prefer a build tag — but I can see the convenience.

(A random drive-by comment without a full review.)

Reply:
@mtrmac Thank you for your feedback!

> This must use a versioned soname.

You're fully right. Adding it.

> Also, are we comfortable with making this a silent fail-open situation? I’m not really, I’d much prefer a build tag — but I can see the convenience.

That's the kind of discussion that I wanted to trigger with this PR :smiley:

Unfortunately there is no way (like an entry in the configuration, for example) to detect if the user intended to use SELinux or not. Therefore I'll add a warning because it's a low-hanging fruit.

Another possibility would be adding a configuration entry like `selinux` defaulting to `true`. The an error would be yielded if the library is missing with the hint that the configuration entry should be changed if the absence of `libselinux` is on purpose.

I find that the second option has the best usability (less possibilities for unexpected behavior). But it also involves wider changes on the code than the 1st one, because it involves changes in `containers/storage`.

Any opinion on this?

Reply:
> Unfortunately there is no way (like an entry in the configuration, for example) to detect if the user intended to use SELinux or not. 

I don’t think we need to answer whether the user _intended to use_ SELinux; The thing _this implementation_ needs to know first, and the thing that is the most painful point for your use, is knowing whether `libselinux.so.1` _is available_ on the target system, whether or not the user enabled it:
- if it is not available, the implementation can’t link to it.
- if it is available, I think the implementation should _always_ link to it, and perhaps use the `libselinux`-determined “is SELinux enabled” configuration value to change its functionality (as the code already does). That’s the definite statement of user intent; we shouldn’t need an extra configuration option.

And “is `libselinux.so.1` expected to be available on the target system” is basically a build tag.  The downside is that every library linking to c/image (that cares about non-SELinux systems) will need to have a `Makefile` possibly setting the build tag.

Reply:
> I don’t think we need to answer whether the user intended to use SELinux; The thing this implementation needs to know first, and the thing that is the most painful point for your use, is knowing whether libselinux.so.1 is available on the target system, whether or not the user enabled it:
> 
> if it is not available, the implementation can’t link to it.
> if it is available, I think the implementation should always link to it, and perhaps use the libselinux-determined “is SELinux enabled” configuration value to change its functionality (as the code already does). That’s the definite statement of user intent; we shouldn’t need an extra configuration option.

I agree with the conditional that you are stating. But currently the decision "is SELinux enabled" is being taken at build-time: if buildtag `selinux` is not enabled, then `vendor/github.com/opencontainers/selinux/go-selinux/selinux_stub.go` gets used and `selinux.GetEnabled() ` returns false. So it doesn't have anything to do with the user intent, it's the builder/packager intent.

That's exactly the point of my patch, avoiding that *implementation needs to know first*.
1. With the current code, SELinux usage is expected if [it was built with support for it](https://github.com/containers/image/pull/546/files#diff-952ad9dc01673ce7143c20dd5808adf7R527). Which is a build-time decision of the builder/packager.
1. With my patch and additional condition is needed to get `containers/image` using SELinux: [`libselinux` is available in the system](https://github.com/containers/image/pull/546/files#diff-952ad9dc01673ce7143c20dd5808adf7R524). Which is a run-time decision of the user.

> And “is libselinux.so.1 expected to be available on the target system” is basically a build tag. The downside is that every library linking to c/image (that cares about non-SELinux systems) will need to have a Makefile possibly setting the build tag.

This sentence is showing exactly what I mean, with the current code it's the builder/packager decision to use SELinux or not, the user's opinion doesn't play a role. But OStree works like a charm without SELinux, so why making it a MUST on build-time? Why taking the decision of using it or not from the user?

Another possibility is providing different packages with and without SELinux support, but I find it's an ugly and confusing solution. If this repository enforces the decision on build-time, then those distributions and integrators not providing SELinux per default and not wanting to enforce their users to use it will just simply patch it out.

Reply:
@mtrmac the **main goal** of this PR is enabling the usage of **OStree without SELinux** without having to patch `ostree/ostree_dest.go` on build-time.

The goal is not really letting the user take the decision. In the end the decision if SELinux is in place or not is mostly being taken by the distributions and users just accept them, although they might change it (e.g. you use Red Hat, you have it, you use Debian, you don't have it). Letting the user decide is more a positive side-effect of the proposed design, than a driver.

In that sense, another possibility to have OStree without SELinux is having another file with the build flag
`// +build !containers_image_ostree_stub,!selinux`,
in that case some refactoring in the code of `ostree/ostree_dest.go` would be needed to avoid code duplication. It's probably the easiest way to approach it without letting the user decide him-/herself.

Reply:
I’m sorry about returning to this so late; I was away for most of December.

> But currently the decision "is SELinux enabled" is being taken at build-time: if buildtag `selinux` is not enabled, then `vendor/github.com/opencontainers/selinux/go-selinux/selinux_stub.go` gets used and `selinux.GetEnabled() ` returns false. So it doesn't have anything to do with the user intent, it's the builder/packager intent.

Sure, it is a builder/packager decision whether to have `libselinux.so` on the target system. The hypothetical user is not expected to rebuild any binaries; the user has `/etc/selinux/config` to express the desired configuration.

> 1. With the current code, SELinux usage is expected if [it was built with support for it](https://github.com/containers/image/pull/546/files#diff-952ad9dc01673ce7143c20dd5808adf7R527). Which is a build-time decision of the builder/packager.
> 2. With my patch and additional condition is needed to get `containers/image` using SELinux: [`libselinux` is available in the system](https://github.com/containers/image/pull/546/files#diff-952ad9dc01673ce7143c20dd5808adf7R524). Which is a run-time decision of the user.

Users are, in general, _not_ supposed to be randomly removing libraries from the system! “Configuration by removing a library” is not how most systems work.

Silently disabling SELinux operation just because a library is missing, or unreadable (perhaps due to a FS error) introduces brittleness into a critical security system, and it seems like something that should be avoided if possible:

Even if you are building an embedded system or a special-purpose system, you just can’t general be pulling binaries from a `libselinux`-dependent OS and expect them to work in a different environment.  From a quick check, all of the following (just a small subset!) directly links to `libselinux` (on a Fedora system): OpenSSH, PAM, shadow-utils, DBus, `sed`, `tar`, OpenVPN, `device-mapper`, `sudo`, `libmount`, `httpd`, `iproute` — as well as, to the point here, `GLib2` and `libostree`.

So, you already need a build system, or a configurable OS with build options, that allow SELinux-less builds of any of the above that your embedded system uses — and `c/image` and/or its users can likewise be a part of that build system with an option that has a SELinux-less build.

Overall, it seems quite possible to make the dependency on `libselinux` a build-time configuration option; and when this is both desirable and possible, why not do it that way?

 
> > And “is libselinux.so.1 expected to be available on the target system” is basically a build tag. The downside is that every library linking to c/image (that cares about non-SELinux systems) will need to have a Makefile possibly setting the build tag.
> 
> This sentence is showing exactly what I mean, with the current code it's the builder/packager decision to use SELinux or not, the user's opinion doesn't play a role. But OStree works like a charm without SELinux, so why making it a MUST on build-time? Why taking the decision of using it or not from the user?

OSTree _does not_ “work like a charm without SELinux” in the way this PR proposes; either you have a `--with-selinux` build or a `--without-selinux` build.  You can’t just take a `--with-selinux` build of OSTree, put it on a system where `libselinux`; it won’t work.

Reply:
> Silently disabling SELinux operation just because a library is missing, or unreadable (perhaps due to a FS error) introduces brittleness into a critical security system, and it seems like something that should be avoided if possible:

… To be fair, this does not quite apply here because, surprisingly enough, `opencontainers/selinux/go-selinux` does not actually link to `libselinux`, so the current implementation (where `go-selinux` first decides whether SELinux is enabled, and then this code can hard-fail if `libselinux` is needed and missing) is not brittle that way.

Still, the `dlopen`s are overall more work than a build tag, and the wider ecosystem needs an “build with `libselinux` switch _anyway_”, so using `dlopen` does not seem worthwhile.